# WeRateDogs Twitter Data Wrangling and Analysis Project
DECI L3 FINAL PROJECT 👨🏻‍🎓

Made by Mohamed Elsayed Zaky 😁

## Prepration 🍰

In [1]:
# Main Libraires
import pandas as pd
import numpy as np
import requests
# Twitter API
import tweepy
import json
# to read tsv file after downlading it using requests
import io
# .env for API keys
from dotenv import load_dotenv
import os

## 1️⃣ Gathering Data 🕵🏻‍♂️

1. Reading `twitter-archive-enhanced.csv` file

In [2]:
tae = pd.read_csv("twitter-archive-enhanced.csv")

2. Reading the `image_predictions.tsv` file

In [3]:
# the url of the image_predictions.tsv file
image_predictions_url = "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
# import it in pandas
image_predictions = requests.get(image_predictions_url)

# Using IO library convert the tsv file to file like object to be able to read
content = image_predictions.content.decode('utf-8')
file_like_object = io.StringIO(content)

In [4]:
nlp = pd.read_csv(file_like_object, sep="\t")

3. Setup Twitter API for addtional data

Note:
- **<div style="color: #e91e63">Because Twitter API Free Plan doesn't offer get_status so I am forced to use the "tweets_json.txt" file in the classroom</div>**

In [5]:
# Gather Three colums `tweet_id`, `retweet_count`, `favorite_count`
tweet_id = []
retweet_count = []
favorite_count = []

# Read twets_json.txt
with open("tweets_json.txt", "r") as file:
  for line in file:
    # Read line as dict
    data = json.loads(line)
    # append data to the list
    tweet_id.append(data['id'])
    retweet_count.append(data['retweet_count'])
    favorite_count.append(data['favorite_count'])
    
tweets_data = {
  "Tweet id": tweet_id,
  "Retweet count": retweet_count,
  "Favorite count": favorite_count,
}

twitter_df = pd.DataFrame(tweets_data)

## 2️⃣ Assessing 👨🏻‍💻

**<h3 style="border-left: 5px solid #673ab7; padding: 0 20px">First: Twitter Archived Enhanced 🚀<h3>**

Let's see the head of the dataset 🕵🏻‍♀️

In [6]:
tae

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

Hmmm! `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` and `retweeted_status_timestamp`  seems to be the majority is null let's check that!

In [7]:
tae['in_reply_to_status_id'].isna().value_counts()

in_reply_to_status_id
True     2278
False      78
Name: count, dtype: int64

In [8]:
tae.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        1611 non-null   object 
 13  doggo                       97 no

There are about 181 retweet let's check that

In [9]:
tae[~tae["retweeted_status_id"].isna()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
32    886054160059072513                    NaN                  NaN   
36    885311592912609280                    NaN                  NaN   
68    879130579576475649                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
...                  ...                    ...                  ...   
1023  746521445350707200                    NaN                  NaN   
1043  743835915802583040                    NaN                  NaN   
1242  711998809858043904                    NaN                  NaN   
2259  667550904950915073                    NaN                  NaN   
2260  667550882905632768                    NaN                  NaN   

                      timestamp  \
19    2017-07-21 01:02:36 +0000   
32    2017-07-15 02:45:48 +0000   
36    2017-07-13 01:35:06 +0000   
68    2017-06-26 00:13:58 +0000   
73    2017-06-24 00:09:53 +0000   
...                         ...   
1023  2016-06-25 01:52:36 +0000   
1043  2016-06-17 16:01:16 +0000   
1242  2016-03-21 19:31:59 +0000   
2259  2015-11-20 03:51:52 +0000   
2260  2015-11-20 03:51:47 +0000   

                                                 source  \
19    <a href="http://twitter.com/download/iphone" r...   
32    <a href="http://twitter.com/download/iphone" r...   
36    <a href="http://twitter.com/download/iphone" r...   
68    <a href="http://twitter.com/download/iphone" r...   
73    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
1023  <a href="http://twitter.com/download/iphone" r...   
1043  <a href="http://twitter.com/download/iphone" r...   
1242  <a href="http://twitter.com/download/iphone" r...   
2259  <a href="http://twitter.com" rel="nofollow">Tw...   
2260  <a href="http://twitter.com" rel="nofollow">Tw...   

                                                   text  retweeted_status_id  \
19    RT @dog_rates: This is Canela. She attempted s...         8.874740e+17   
32    RT @Athletics: 12/10 #BATP https://t.co/WxwJmv...         8.860537e+17   
36    RT @dog_rates: This is Lilly. She just paralle...         8.305833e+17   
68    RT @dog_rates: This is Emmy. She was adopted t...         8.780576e+17   
73    RT @dog_rates: Meet Shadow. In an attempt to r...         8.782815e+17   
...                                                 ...                  ...   
1023  RT @dog_rates: This is Shaggy. He knows exactl...         6.678667e+17   
1043  RT @dog_rates: Extremely intelligent dog here....         6.671383e+17   
1242  RT @twitter: @dog_rates Awesome Tweet! 12/10. ...         7.119983e+17   
2259  RT @dogratingrating: Exceptional talent. Origi...         6.675487e+17   
2260  RT @dogratingrating: Unoriginal idea. Blatant ...         6.675484e+17   

      retweeted_status_user_id retweeted_status_timestamp  \
19                4.196984e+09  2017-07-19 00:47:34 +0000   
32                1.960740e+07  2017-07-15 02:44:07 +0000   
36                4.196984e+09  2017-02-12 01:04:29 +0000   
68                4.196984e+09  2017-06-23 01:10:23 +0000   
73                4.196984e+09  2017-06-23 16:00:04 +0000   
...                        ...                        ...   
1023              4.196984e+09  2015-11-21 00:46:50 +0000   
1043              4.196984e+09  2015-11-19 00:32:12 +0000   
1242              7.832140e+05  2016-03-21 19:29:52 +0000   
2259              4.296832e+09  2015-11-20 03:43:06 +0000   
2260              4.296832e+09  2015-11-20 03:41:59 +0000   

                                          expanded_urls  rating_numerator  \
19    https://twitter.com/dog_rates/status/887473957...                13   
32    https://twitter.com/dog_rates/status/886053434...                12   
36    https://twitter.com/dog_rates/status/830583320...                13   
68    https://twitt

Let's also check the repalyed tweets

In [10]:
tae[~tae["in_reply_to_status_id"].isna()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
...                  ...                    ...                  ...   
2038  671550332464455680           6.715449e+17         4.196984e+09   
2149  669684865554620416           6.693544e+17         4.196984e+09   
2169  669353438988365824           6.678065e+17         4.196984e+09   
2189  668967877119254528           6.689207e+17         2.143566e+07   
2298  667070482143944705           6.670655e+17         4.196984e+09   

                      timestamp  \
30    2017-07-15 16:51:35 +0000   
55    2017-07-02 21:58:53 +0000   
64    2017-06-27 12:14:36 +0000   
113   2017-06-02 19:38:25 +0000   
148   2017-05-13 16:15:35 +0000   
...                         ...   
2038  2015-12-01 04:44:10 +0000   
2149  2015-11-26 01:11:28 +0000   
2169  2015-11-25 03:14:30 +0000   
2189  2015-11-24 01:42:25 +0000   
2298  2015-11-18 20:02:51 +0000   

                                                 source  \
30    <a href="http://twitter.com/download/iphone" r...   
55    <a href="http://twitter.com/download/iphone" r...   
64    <a href="http://twitter.com/download/iphone" r...   
113   <a href="http://twitter.com/download/iphone" r...   
148   <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2038  <a href="http://twitter.com/download/iphone" r...   
2149  <a href="http://twitter.com/download/iphone" r...   
2169  <a href="http://twitter.com/download/iphone" r...   
2189  <a href="http://twitter.com/download/iphone" r...   
2298  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
30    @NonWhiteHat @MayhewMayhem omg hello tanner yo...                  NaN   
55    @roushfenway These are good dogs but 17/10 is ...                  NaN   
64                      @RealKentMurphy 14/10 confirmed                  NaN   
113   @ComplicitOwl @ShopWeRateDogs &gt;10/10 is res...                  NaN   
148   @Jack_Septic_Eye I'd need a few more pics to p...                  NaN   
...                                                 ...                  ...   
2038  After 22 minutes of careful deliberation this ...                  NaN   
2149  After countless hours of research and hundreds...                  NaN   
2169  This is Tessa. She is also very pleased after ...                  NaN   
2189                     12/10 good shit Bubka\n@wane15                  NaN   
2298  After much debate this dog is being upgraded t...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
30                         NaN                        NaN   
55                         NaN                        NaN   
64                         NaN                        NaN   
113                        NaN                        NaN   
148                        NaN                        NaN   
...                        ...                        ...   
2038                       NaN                        NaN   
2149                       NaN                        NaN   
2169                       NaN                        NaN   
2189                       NaN                        NaN   
2298                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
30                                                  NaN                12   
55                                                  NaN                17   
64                                                  NaN                14   
113                

There are also 78 rows or entry is a replay

I observed that `expanded_urls` are non except for row number `2169` so let's check that

In [11]:
tae["expanded_urls"][2169]

'https://twitter.com/dog_rates/status/669353438988365824/photo/1'

After check it. It's a replay in a thread method. The @dog_rates account makes two tweets for the same dog in a thread.

So let's check threads

In [12]:
replay_tae = tae[~tae["in_reply_to_status_id"].isna()]
replay_tae["name"].isna().value_counts()

name
True     77
False     1
Name: count, dtype: int64

In [13]:
replay_tae[replay_tae["name"].isna() == False]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2169  669353438988365824           6.678065e+17         4.196984e+09   

                      timestamp  \
2169  2015-11-25 03:14:30 +0000   

                                                 source  \
2169  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2169  This is Tessa. She is also very pleased after ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2169                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2169  https://twitter.com/dog_rates/status/669353438...                10   

      rating_denominator   name doggo floofer pupper puppo  
2169                  10  Tessa   NaN     NaN    NaN   NaN

So if we checked the name it by the name only. As there are some links not working. We will find that `2169` is the only replay we need.

Let's see the `name` column

In [14]:
def print_max_rows(func):
  pd.set_option("display.max_rows", None) # to display max rows
  print(func)
  pd.reset_option("display.max_rows", None) # reset settings

In [15]:
print_max_rows(tae["name"].value_counts())

name
a                 55
Charlie           12
Oliver            11
Cooper            11
Lucy              11
Lola              10
Tucker            10
Penny             10
Bo                 9
Winston            9
the                8
Sadie              8
Daisy              7
Buddy              7
Bailey             7
Toby               7
an                 7
Bella              6
Jack               6
Oscar              6
Rusty              6
Stanley            6
Scout              6
Jax                6
Leo                6
Milo               6
Koda               6
Dave               6
Oakley             5
Larry              5
Sunny              5
Louis              5
Chester            5
George             5
Alfie              5
Finn               5
very               5
Sammy              5
Gus                5
Phil               5
Bentley            5
Sampson            4
Reginald           4
quite              4
Boomer             4
Reggie             4
Loki               4
Dexter  

In [16]:
tae["name"].isna().value_counts()[True]

745

I observed that there are wrong names and all wrong names starts with small letters like: `a`, `the`, `very`, `quite`, `just` and `an`.

So let's check names starts with small and capital letters. And also the NaNs.

In [17]:
tae["name"].fillna(value="nan", inplace=True)
tae["name"].str.istitle().value_counts()

name
True     1496
False     860
Name: count, dtype: int64

Notes:
  - There are about 745 null values.
  - There are also incorret names like `a`, `the`, `very`, `quite`, `just` and `an`.
  - There are 115 rows it's names is not correct.

Also I am wandring what's inside the `text` so let's take a random entry

In [18]:
tae["text"][1500]

"This is Edgar. He's a Sassafras Puggleflash. Nothing satisfies him. Not since the war. 10/10 cheer up pup https://t.co/1NgMb9BTWB"

No need to the last link, it gives me the link of the tweet

**<h3 style="border-left: 5px solid #673ab7; padding: 0 20px">Secound: Image Predection Dataset 🔮<h3>**

In [19]:
nlp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [20]:
nlp.head() # 315

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [21]:
nlp[["p1_dog"	,"p2_dog"	,"p3_dog"]].value_counts()

p1_dog  p2_dog  p3_dog
True    True    True      1243
False   False   False      324
True    True    False      155
False   True    True       101
True    False   True        91
False   False   True        64
        True    False       54
True    False   False       43
Name: count, dtype: int64

Notes:
- There are about 1243 entry or 60% of dataset are predicted to be true by the the models. (Most Likely to be images of dogs) 😉
- There are totally 411 entry are true by two models. (Almost is images of dogs) 🙂
- There are 161 entry are true by one model only. (May be an image of dog) 😕
- 324 entry are totally false (low percentage to be an image of dog) 😟

**<h3 style="border-left: 5px solid #673ab7; padding: 0 20px">Third: Twitter Json 👨🏻‍💻<h3>**

In [22]:
twitter_df.head()

Tweet id  Retweet count  Favorite count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

**<h3 style="border-left: 5px solid #673ab7; padding: 0 20px">Issues in a nutsell 🤓🤓<h3>**

**Quality Issues 🥇**
- `timestamp` column is string rather than datetime object
- `timestamp` must have the date only no need to the time.
- `source` column has the html anchor tage `<a>` with the link. It must be the link only.
- No need to `source` column as it's the same repeated value. No need to visit the settings page.
- There are five colums have more than 70% of it null values.
- `expanded_urls` is just this 'https://twitter.com/dog_rates/status/' with the id of the tweet. no need to it as we have the id.
- `retweeted_status_id`, `retweeted_status_user_id` and `retweeted_status_timestamp` Tells us that there are about 181 entries is just a retweet. we do not need them.
- `name` has about 745 null values
- `name` 115 rows it's names is not correct
- There are about 78 row is an replays
- Some images in Image Predection Dataset are not for dogs.

**Tidness Issues 🧹**:
- The `doggo`, `floofer`, `pupper`, and `puppo` columns are just terms that refers to the dogs 
characteristics. It must be one colum called `dog characteristics`

## 3️⃣ Cleaning 🧼

**<h3 style="border-left: 5px solid #673ab7; padding: 0 20px">First: Twitter Archived Enhanced 🚀<h3>**

In [23]:
cleaned_tae = tae.copy()

In [24]:
# pd.options.mode.chained_assignment = None # to prevent warnings

Let's drop any replays except column `2169`

In [25]:
cleaned_tae['in_reply_to_status_id'][2169] = np.nan # Set the column 2169 to be nan

cleaned_tae = cleaned_tae[cleaned_tae["in_reply_to_status_id"].isna()]

C:\Users\smy\AppData\Local\Temp\ipykernel_12340\285241794.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaned_tae['in_reply_to_status_id'][2169] = np.nan # Set the column 2169 to be nan


Let's drop the columns that we doesn't need it.

In [26]:
cleaned_tae.drop(columns = ['in_reply_to_status_id', 'in_reply_to_user_id', 'source', 'expanded_urls'], inplace=True)

Also let's drop any retweet entry

In [27]:
cleaned_tae = cleaned_tae[cleaned_tae['retweeted_status_id'].isna()]

Then no need to `retweeted_status_id`, `retweeted_status_user_id`, and `retweeted_status_timestamp` columns

In [28]:
cleaned_tae.drop(columns = ['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], inplace=True)

Convert time stamp to datetime and trim the time

In [29]:
cleaned_tae["timestamp"] = pd.to_datetime(cleaned_tae["timestamp"].str.slice(0, 10))

let's clean text column from the last link

In [30]:
cleaned_tae["text"] = cleaned_tae["text"].str[:-24]

Let's clean the `name` column

In [31]:
cleaned_tae[~cleaned_tae["name"].str.istitle()]["name"].value_counts()

name
nan             603
a                55
the               8
an                6
one               4
very              4
quite             3
just              3
CeCe              2
actually          2
getting           2
not               2
his               1
my                1
JD                1
space             1
light             1
life              1
officially        1
by                1
incredibly        1
infuriating       1
old               1
all               1
unacceptable      1
this              1
mad               1
such              1
BeBe              1
DonDon            1
DayZ              1
Name: count, dtype: int64

In [32]:
# need to replace nan values

In [33]:
cleaned_tae

tweet_id  timestamp  \
0     892420643555336193 2017-08-01   
1     892177421306343426 2017-08-01   
2     891815181378084864 2017-07-31   
3     891689557279858688 2017-07-30   
4     891327558926688256 2017-07-29   
...                  ...        ...   
2351  666049248165822465 2015-11-16   
2352  666044226329800704 2015-11-16   
2353  666033412701032449 2015-11-15   
2354  666029285002620928 2015-11-15   
2355  666020888022790149 2015-11-15   

                                                   text  rating_numerator  \
0     This is Phineas. He's a mystical boy. Only eve...                13   
1     This is Tilly. She's just checking pup on you....                13   
2     This is Archie. He is a rare Norwegian Pouncin...                12   
3     This is Darla. She commenced a snooze mid meal...                13   
4     This is Franklin. He would like you to stop ca...                12   
...                                                 ...               ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                 5   
2352  This is a purebred Piers Morgan. Loves to Netf...                 6   
2353  Here is a very happy pup. Big fan of well-main...                 9   
2354  This is a western brown Mitsubishi terrier. Up...                 7   
2355  Here we have a Japanese Irish Setter. Lost eye...                 8   

      rating_denominator      name doggo floofer pupper puppo  
0                     10   Phineas   NaN     NaN    NaN   NaN  
1                     10     Tilly   NaN     NaN    NaN   NaN  
2                     10    Archie   NaN     NaN    NaN   NaN  
3                     10     Darla   NaN     NaN    NaN   NaN  
4                     10  Franklin   NaN     NaN    NaN   NaN  
...                  ...       ...   ...     ...    ...   ...  
2351                  10       nan   NaN     NaN    NaN   NaN  
2352                  10         a   NaN     NaN    NaN   NaN  
2353                  10         a   NaN     NaN    NaN   NaN  
2354                  10         a   NaN     NaN    NaN   NaN  
2355                  10       nan   NaN     NaN    NaN   NaN  

[2098 rows x 10 columns]

First let's make a copy of tae dataframe 📁

## 4️⃣ Storing data 🧠

## 5️⃣ Analyzing, and visualizing data 📊

## 6️⃣ Reporting 👨🏻‍🏫📃